### PR TITLE
Feature/1.21.4

### PR DESCRIPTION
**Pull request by onebeastchris**
Opening a PR to track to-do's.

TODO:

Left to implement:
- [x] actually load generated default data components
- [x] new blocks/items
- [x] new boats
- [x] creaking
- [x] new sounds/particles
- [ ] block breaking changes -> Bedrock can see their own progress fine, but not progress by other players :/
- [x] send ServerboundPlayerLoadedPacket
- [x] handle new repair item / tool components
- [x] yeet tooltier/repair items
- [x] new item/block tags
~~- [ ] repair cost is a component too! https://github.com/GeyserMC/MCProtocolLib/blob/9332151a49e38a9705998609075a8a8c8a887903/protocol/src/main/java/org/geysermc/mcprotocollib/protocol/data/game/item/component/DataComponentType.java#L41~~ will deal with in separate PR

Other:
- [x] Update/target Spigot adapter

Needs testing:
- [x] test block/entity picking

### CODE REVIEW

File: core/src/main/java/org/geysermc/geyser/entity/type/living/monster/CreakingEntity.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        } else {
+            propertyManager.add("minecraft:creaking_state", "neutral");
+        }
+        GeyserImpl.getInstance().getLogger().warning("set active; " + booleanEntityMetadata.toString());
+    }
+
+    public void setIsTearingDown(EntityMetadata<Boolean,? extends MetadataType<Boolean>> booleanEntityMetadata) {
+        GeyserImpl.getInstance().getLogger().warning("set isTearingDown; " + booleanEntityMetadata.toString());
+        if (booleanEntityMetadata.getValue()) {
+            propertyManager.add("minecraft:creaking_state", "crumbling");
```
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/entity/type/living/monster/CreakingEntity.java:103]: We should probably make this a String constant.

File: core/src/main/java/org/geysermc/geyser/inventory/updater/AnvilInventoryUpdater.java [hunk 2]
```diff
@@ -395,8 +397,14 @@ private boolean isCombining(GeyserItemStack input, GeyserItemStack material) {
...
 
-    private boolean isRepairing(GeyserItemStack input, GeyserItemStack material) {
-        return input.asItem().isValidRepairItem(material.asItem());
+    private boolean isRepairing(GeyserItemStack input, GeyserItemStack material, GeyserSession session) {
+        HolderSet repairable = input.getComponent(DataComponentType.REPAIRABLE);
+        if (repairable == null) {
+            return false;
+        }
+
+        GeyserHolderSet<Item> set = GeyserHolderSet.convertHolderSet(JavaRegistries.ITEM, repairable);
```
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/inventory/updater/AnvilInventoryUpdater.java:406]: In this case, might just be better to have an #is(HolderSet, Item) method so we don't have to create a new object?

File: core/src/main/java/org/geysermc/geyser/translator/inventory/LoomInventoryTranslator.java [hunk 3]
```diff
@@ -156,16 +154,11 @@ public ItemStackResponse translateSpecialRequest(GeyserSession session, Inventor
...
         inputCopy.setNetId(session.getNextItemNetId());
-        // Add the pattern manually, for better item synchronization
-        if (inputCopy.getComponents() == null) {
-            inputCopy.setComponents(new DataComponents(new HashMap<>()));
-        }
-
         BannerPatternLayer bannerPatternLayer = BannerItem.getJavaBannerPattern(session, pattern); // TODO
         if (bannerPatternLayer != null) {
-            List<BannerPatternLayer> patternsList = inputCopy.getComponents().getOrDefault(DataComponentType.BANNER_PATTERNS, new ArrayList<>());
+            List<BannerPatternLayer> patternsList = inputCopy.getComponentOrFallback(DataComponentType.BANNER_PATTERNS, new ArrayList<>());
```
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/translator/inventory/LoomInventoryTranslator.java:159]: Not sure how we missed this one, but every time this is called, a new ArrayList will be created.

Comment (onebeastchris) [reply core/src/main/java/org/geysermc/geyser/translator/inventory/LoomInventoryTranslator.java]: Fixed in https://github.com/GeyserMC/Geyser/commit/c298061d4dea3569a285b425bc4442d9e8445adc

**Comment by TutorialMan7727** [comment]
How long until this pr is prod ready

**Comment by onebeastchris** [comment]
You can already use it fine - the only thing that's currently still not quite there are block breaking animations for other players, but those are only visual anyways.

**Comment by DivineJimmi** [comment]
Anyone able to make a .jar of this?

**Comment by eclipseisoffline** [comment]
To access build artifacts, you can click on "Checks" at the top of this pull request, then click on "Artifacts"

PR1

**Pull request merged**